### ISSUE TITLE
Feature request: QStackedWidget-based wizard-like functionality

### ISSUE THREAD

**Issue by scott-maddox** [enthought/traitsui#700]
There have a been a couple of projects where having some built-in wizard-like functionality would be very helpful. By wizard-like, I mean having a progression of views that get pushed onto a metaphorical stack when you click "next" (or whatever), or popped off when you click "back", in which the state of the past views are maintained and recovered when you click "back".

A critical aspect of this is that it should be possible for each view that gets pushed on to the stack to be generated dynamically when "next" (or whatever) is pushed, and that different views can be generated depending on the interactions that have occurred on the current view.

I've implemented this at the pyqt layer using QStackedWidget. It's not clear if or how such a concept could be implemented or what API would be exposed for first-class support in TraitsUI. But I do think it could be useful for many projects.

**Comment by corranwebster** [comment]
TraitsUI has the concept of `kind='wizard'` for Views.  It should work for the Wx backend, but was never properly implemented for the Qt backend.  If it were to be implemented in the Qt backend, it would probably make sense to base it off of the `QWizard` widget (https://doc.qt.io/archives/qt-4.8/qwizard.html) rather than a `QStackedWidget`.

**Comment by kSkip** [comment]
We previously tried using the `QWizard` for our case, but it is meant to be used as a modal dialog with a very specific api. It doesn't size/resize properly if you mount in another window. However, the `QStackedWidget` is a simpler way of paging and addressing views that need to be embedded without the main window.

I guess alternatively, you could use notebook tabs for this, but that is not scalable beyond the size of your labels given a certain window size.

IMO both should be supported.

**Comment by rkern** [comment]
One of the reasons that `kind='wizard'` has withered on the vine is that it was hard to come up with an API that is actually less work to use than just doing it yourself but is still flexible enough to account for all of the use cases. It's much easier if there's a static sequence of pages, but if you need to dynamically generate the next page, working around whatever general API you make is probably no better than just using an `InstanceEditor` and swapping in the right instance using whatever logic you need at the time. Don't even need `QStackedWidget`.

**Comment by scott-maddox** [comment]
That sounds significantly more complicated than `QStackedWidget`. I tried doing something similar (except at the QT level), and it turned into a nightmare as soon as I wanted more than 2 pages in the stack. Switching to `QStackedWidget` made it much more streamlined.

**Comment by kSkip** [comment]
The `QStackedWidget` can manage the complexity of ordering and indexing. Using an `InstanceEditor` requires you to manually manage view changes on push/pop/index operations.

I'm not saying it definitely does fit into TraitsUI, but It seems similar to how the `EnumEditor` can populate it's choices by syncing with a `List` using `name='list_trait'` or how the TabularEditor can sync the selected item with a trait attribute.

The signals and slots are a 1-to-1 mapping of this logic
Example:
```
Item('views', index='view_idx')
```

When `views` is appended `addWidget(QWidget *widget)` is called. When `view_idx` is changed `setCurrentIndex(int index)` is called.

And so on. It just doesn't seem that different from the `EnumEditor` behavior.

**Comment by rkern** [comment]
Not sure what you were doing at the Qt level that made it a nightmare, but ultimately, `QStackedWidget` is just a list with a current index, and managing a `list` of instances and an index integer shouldn't be too stressful. If it's the reparenting that was causing you problems, that's not an issue with using `InstanceEditor` in this way. Those details are taken care of for you. The static case looks something like this:

```python
from traits.api import *
from traitsui.api import *


class StaticWizzard(HasTraits):
    pages = List()
    page_index = Int(0)
    current_page = Instance(HasTraits)
    next_button = Button('>>')
    prev_button = Button('<<')

    traits_view = View(
        VGroup(
            HGroup(
                UItem('prev_button', enabled_when='page_index > 0'),
                spring,
                UItem('next_button', enabled_when='page_index < len(pages) - 1'),
            ),
            UItem('current_page', style='custom'),
        ),
        resizable=True,
    )

    def _next_button_changed(self):
        self.page_index += 1

    def _prev_button_changed(self):
        self.page_index -= 1

    def _page_index_changed(self):
        self.current_page = self.pages[self.page_index]

    def _current_page_default(self):
        return self.pages[0]


class Page(HasTraits):
    text = Str()


w = StaticWizzard(
    pages=[Page(text=f'Page {i}') for i in range(5)],
)
w.configure_traits()
```

The dynamic case shouldn't be much harder. Do whatever you do with `QStackedWidget` and replace `addWidget(widget)` with `self.pages.append(page)`

The real complication comes in handling just the right amount dynamism that you want, but `QStackedWidget` doesn't handle that for you. For example, you say that you need to be able to generate the next page dynamically based on the current results. When the user goes back and changes something, do the following pages get a chance to be replaced? And if the user goes back and doesn't make a change (or makes a change that doesn't require a dynamic replacement), do they keep the following pages just as they left them?

**Comment by scott-maddox** [comment]
Thanks, Robert, something based on that might work. When I have some time, I'll test it out.

I did have to add some logic in the raw QT implementation to resize the window appropriately on the view change, to keep things properly proportioned. And the buttons change depending on the view (e.g. instead of Back button, an Abort button, which gets disabled once clicked until abort is complete). I'll have to see if/how those complications would fit into this approach.

For simpler Wizards, I can definitely see something like your example working, as long as variation in view sizes doesn't cause problems.

Maybe we could add a HowTo section on building a wizard to the documentation, based on your code?

**Comment by rkern** [comment]
Yeah, it's those complications that mean that building a wizard framework might not be indicated. Just navigating the sufficiently-flexible framework can be as complicated as assembling the more primitive components yourself.

This kind of thing is going to have to come up from the individual projects rather than being designed in the abstract. Do share your solution with others, and if you find that you've been able to deploy the same general code for 3 or so projects, then we can think about exposing that general code in `traitsui`.

**Comment by scott-maddox** [comment]
A word of caution: my implementation is by no means general. It's very specific to our particular use case. I agree that trying to design something that could handle all these complexities in a general way would be difficult, and potentially not worth the effort.

But perhaps having something like a HowTo section in the documentation on building a wizard, based on your code, would have a much smaller scope, and could be a valuable place to point new developers who want that kind of a workflow?
